src/vm/synch.cpp

```diff
@@ -721,7 +719,7 @@ bool CLRLifoSemaphore::Wait(DWORD timeoutMs)
     _ASSERTE(m_handle != nullptr);
 
     // Acquire the semaphore or register as a waiter
-    Counts counts = m_counts.VolatileLoad();
```

Comment:
Why is it that the volatile load isn't necessary in these cases?  I think it'd be worth explaining the reasoning in the commit description.

Reply:
The main reason is that there isn't any reason why they are necessary. I had added them initially out of pattern and could have left them out. There was some concern before that without a volatile load a compiler could replace local uses with a memory load and that would change the meaning of what is intended, but that doesn't apply to these kind of loops because it would be incorrect to do so, the concern may apply to loops like the following:

```c#
while(true)
{
    Counts counts = m_counts;
    if(InterlockedCompareExchange(&m_counts, counts + 1, counts) == counts)
        break;
}
```

Where if the 3rd argument is replaced by the compiler with m_counts, it may yield an incorrect result. It's an invalid optimization but anyway that can't be done in the type of loops used in this code because the initial value used for subsequent loop iterations is the result of the compare-exchange operation and not the memory location.

Reply:
If the first loop iteration is unrolled in the current code it would look similar to the code above where the concern could apply. However, similar volatiles were removed in Monitor a long time ago (in EnterEpilogHelper and other places for example, where they explicitly do non-volatile loads in loops like above). So I don't think that compiler issue is something we need to worry about anymore.